src/funnels/SwapDepositor.sol

```diff
@@ -0,0 +1,222 @@
+    event File (bytes32 indexed what, address gem, uint256 data);
+
+    constructor(address roles_, bytes32 ilk_, address uniV3Factory_) {
+        roles = RolesLike(roles_);
+        ilk = ilk_;
+        uniV3Factory = uniV3Factory_;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(roles.canCall(ilk, msg.sender, address(this), msg.sig) || wards[msg.sender] == 1, "SwapDepositor/not-authorized");
+        _;
+    }
+
+    function rely(address usr) external auth { wards[usr] = 1; emit Rely(usr); }
+    function deny(address usr) external auth { wards[usr] = 0; emit Deny(usr); }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "swapper")    swapper   = data;
+        else if (what == "depositor")  depositor = data;
+        else revert("SwapDepositor/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address gem, uint256 data) external auth {
+        if      (what == "gap")  gaps[gem] = data;
+        else revert("SwapDepositor/file-unrecognized-param");
+        emit File(what, gem, data);
+    }
+
+
+    // https://github.com/Uniswap/v3-periphery/blob/464a8a49611272f7349c970e0fadb7ec1d3c1086/contracts/libraries/PoolAddress.sol#L33
+    function getPoolAddress(address gem0, address gem1, uint24 fee) internal view returns (address pool) {
+        pool = address(uint160(uint256(keccak256(abi.encodePacked(
+            hex'ff',
+            uniV3Factory,
+            keccak256(abi.encode(gem0, gem1, fee)),
+            bytes32(0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54) // POOL_INIT_CODE_HASH
+         )))));
+    }
+
+    struct SwapDepositParams {
+        address gem0;
+        address gem1;
+        uint256 amt0;
+        uint256 amt1;
+        uint256 minAmt0;
+        uint256 minAmt1;
+        uint256 minSwappedOut;
+        uint24 fee;
+        int24 tickLower;
+        int24 tickUpper;
+        address swapperCallee;
+        bytes swapperData;
+    }
+
+    function sortDepositedTokens(SwapDepositParams memory p) internal pure returns (SwapDepositParams memory) {
+        (p.gem0, p.gem1, p.amt0, p.amt1, p.minAmt0, p.minAmt1) 
+            = p.gem0 < p.gem1 ? (p.gem0, p.gem1, p.amt0, p.amt1, p.minAmt0, p.minAmt1) : (p.gem1, p.gem0, p.amt1, p.amt0, p.minAmt1, p.minAmt0);
+        return p;
+    }
+
+    function getConvertedAmounts(SwapDepositParams memory p) internal view returns (uint256 amt0_1, uint256 amt1_0) {
+        (uint160 sqrtPriceX96,,,,,,) = UniV3PoolLike(getPoolAddress(p.gem0, p.gem1, p.fee)).slot0();
+        uint256 p0 = uint256(sqrtPriceX96) * uint256(sqrtPriceX96) * WAD / uint256(2 ** 192);
+        amt0_1 = p.amt0 * p0 / WAD;
+        amt1_0 = p.amt1 * WAD / p0;
+    }
+
+    function swapBeforeDeposit(SwapDepositParams memory p) internal returns (uint256 amt0, uint256 amt1) {
+        (uint256 amt0_1, uint256 amt1_0) = getConvertedAmounts(p);
```

Comment:
I think the flexibility of processing any amt of the two tokens (and not just assuming we swap `lot` of one to the other and then deposit like the flapper) is a nice idea, but there's a problem that `minAmt0`, `minAmt1`, `minSwappedOut` can't be predetermined easily since they depend on the price. I think in general we might give too much flexibility here and would be hard to account for all outcomes/manipulation possibilities.

Reply:
Resolving as we'll do an automated depositor only.